ignite/internal/tools/gen-mig-docs/main.go

```diff
@@ -0,0 +1,630 @@
+	// A temporary directory is created to clone ignite cli repository and build it
+	tmpdir, err := os.MkdirTemp("", ".migdoc")
+	defer os.RemoveAll(tmpdir)
+	if err != nil {
+		return err
+	}
+	logger.Println("Created temporary directory:", tmpdir)
+
+	logger.Println("Cloning", igniteCliRepository)
+	repoDir := filepath.Join(tmpdir, igniteRepoPath)
+	repo, err := git.PlainClone(repoDir, false, &git.CloneOptions{
+		URL: igniteCliRepository,
+	})
+	if err != nil {
+		return err
+	}
+
+	versions, err := getRepositoryVersionTags(repo)
+	if err != nil {
+		return err
+	}
+
+	fromVer, toVer, err = validateVersionRange(fromVer, toVer, versions)
+
+	logger.Printf("Generating migration document for %s->%s\n", fromVer, toVer)
+
+	wt, err := repo.Worktree()
+	if err != nil {
+		return errors.Wrap(err, "failed to get worktree")
+	}
+
+	// Run scaffolds for fromVer and toVer
+	fromVerDir := filepath.Join(tmpdir, fromVer.Original())
+	err = runScaffoldsForVersion(wt, repoDir, fromVerDir, fromVer)
+	if err != nil {
+		return errors.Wrapf(err, "failed to run scaffolds for tag %s", fromVer)
+	}
+	toVerDir := filepath.Join(tmpdir, toVer.Original())
+	err = runScaffoldsForVersion(wt, repoDir, toVerDir, toVer)
+	if err != nil {
+		return errors.Wrapf(err, "failed to run scaffolds for tag %s", toVer)
+	}
+
+	// Run diff between two directories
+	logger.Println("Generating diff...")
+	diffMap, err := calculateDiff(fromVerDir, toVerDir)
+	if err != nil {
+		return err
+	}
+	err = subtractBaseDiffs(diffMap)
+	if err != nil {
+		return err
+	}
+
+	outputDir := fmt.Sprintf("migdoc-%s-%s", fromVer, toVer)
+	err = os.MkdirAll(outputDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = saveDiffMap(diffMap, outputDir)
+	if err != nil {
+		return err
+	}
+	logger.Println("Migration document generated successfully at", outputDir)
+
+	return nil
+}
+
+func getRepositoryVersionTags(repo *git.Repository) (semver.Collection, error) {
+	tags, err := repo.Tags()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get tags")
+	}
+
+	// Iterate over all tags in the repository and pick valid semver tags
+	var versions semver.Collection
+	err = tags.ForEach(func(ref *plumbing.Reference) error {
+		name := ref.Name()
+		if name.IsTag() {
+			ver, err := semver.NewVersion(name.Short())
+			if err != nil {
+				// Do nothing as it's not a semver tag
+				return nil
+			}
+			versions = append(versions, ver)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to iterate over tags")
+	}
+
+	sort.Sort(versions)
+
+	return versions, nil
+}
+
+func validateVersionRange(fromVer, toVer *semver.Version, versions semver.Collection) (*semver.Version, *semver.Version, error) {
+	// Unable to generate migration document if there are less than two releases!
+	if versions.Len() < 2 {
+		return nil, nil, errors.New("At least two semver tags are required")
+	}
+
+	// Replace fromVer and toVer with equivalent semver tags from versions
+	if fromVer != nil {
```

Comment:
why we need these validations?

Reply:
don't pass the from will break the logic anyway

Reply:
This is for when we pass `from` or `to` versions manually so it has to check if a git tag really exists. Also it has to check if the `to` value is greater than `from` according to semver standard.